### PR TITLE
[Index] Add SymbolKind -> indexstore_symbol_kind_t conversion for tem…

### PR DESCRIPTION
**Pull request by jkorous-apple**
…plate parameters

### CODE REVIEW

**Comment by akyrtzi** [comment]
I generally would like for `SymbolKind` to be more generic and language independent when possible, and use `SymbolSubKind` for language specific types.

What do you think about proposing in clang upstream to add a `SymbolKind::GenericParameter` kind and add
```
CXXTemplateTypeParm,
CXXTemplateTemplateParm,
CXXNonTypeTemplateParm,
```
as `SymbolSubKind`s ?

**Comment by jkorous-apple** [comment]
I'd say let's address this first so we get rid of the warning:
```
.../llvm-project/clang/lib/Index/IndexDataStoreUtils.cpp:286:11: warning: enumeration values 'TemplateTypeParm', 'TemplateTemplateParm', and 'NonTypeTemplateParm' not handled in switch [-Wswitch]
```

Couple thoughts re: splitting `SymbolKind`/`SymbolSubKind` into more generic/more language specific kinds.
- The enums are mostly used the way you suggest but not completely.
- used both in libclang and clangd
- There might be less intrusive ways (less maintenance friction) how to do this?

**Comment by akyrtzi** [comment]
> I'd say let's address this first so we get rid of the warning:

While the clang C++ code can change in any way anytime, we have to be more careful when changing the API interface of libIndexStore because we need to keep it backwards compatible.

In this case I don't think it is proper to assign these new symbol kinds as `INDEXSTORE_SYMBOL_KIND_PARAMETER`. Ideally symbol kinds should be reflected via separate `INDEXSTORE_SYMBOL_KIND*` entries, but in this case, since it is not clear to me that the best approach is for all these 3 to be symbol kinds, we should not be adding `INDEXSTORE_SYMBOL_KIND*` entries that may go away later.

If you'd like in the short term to address the warning, assign the new kinds to `INDEXSTORE_SYMBOL_KIND_UNKNOWN` and add a `FIXME` to revisit whether they should be symbol kinds or subkinds and get to a consensus, before we expose them via the `INDEXSTORE_*` API.

File: clang/lib/Index/IndexDataStoreUtils.cpp [hunk 1]
```diff
@@ -285,6 +285,9 @@ SymbolRoleSet index::getSymbolRoles(uint64_t Roles) {
 indexstore_symbol_kind_t index::getIndexStoreKind(SymbolKind K) {
   switch (K) {
   case SymbolKind::Unknown:
+  case SymbolKind::TemplateTypeParm:
+  case SymbolKind::TemplateTemplateParm:
+  case SymbolKind::NonTypeTemplateParm:
     return INDEXSTORE_SYMBOL_KIND_UNKNOWN;
```
Comment (akyrtzi) [inline clang/lib/Index/IndexDataStoreUtils.cpp:304]: Could you also add a `FIXME` comment to indicate that we should figure out the best way to expose this info?

**Review by akyrtzi — APPROVED** [review]

**Pull request merged**